### DISCUSSION TITLE
mismatched input 's' expecting ID

### DISCUSSION THREAD

**Discussion by lexasub** [antlr/antlr4#3882] [General]
hello, i get error:mismatched input 'import' expecting 'import'. 
i use this code: CharStream stream = CharStreams.fromFileName("./test", StandardCharsets.US_ASCII);
        langosLexer lexer = new langosLexer(stream);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        langosWithoutSyntaxParser parser = new langosWithoutSyntaxParser(tokens);
        MylangosWithoutSyntaxVisitor visitor = new MylangosWithoutSyntaxVisitor();
        System.out.print(visitor.visitEntry_point(parser.entry_point()));
//https://github.com/lexasub/langA/blob/master/src/org/lexasub/langosThirdTryWithoutPromise/IO.java
my grammatic:
https://github.com/lexasub/langA/blob/master/langosWithoutSyntax.g4

**Comment by parrt** [comment]
You need to look at what token type is actually being matched; no doubt there is a duplicate rule to match it important

**Reply by lexasub** [reply]
can I please elaborate? I didn't understand you

**Reply by lexasub** [reply]
*may you say about it more?

**Comment by parrt** [comment]
sure. First i suggest using the ANTLR plugin in intellij. Next you likely have a rule set in this order:

```
ID : [a-z]+ ;
IMPORT : 'import' ;
```

Then the parser will look for IMPORT but see ID.

**Reply by lexasub** [reply]
in ANTLR plugin in intellij parse tree on"import a;" is normal.
//temporary i reduce rule entry_point->entry_point : import_ EOF;
             entry_point
import_              eof
IMPORT ID:a SEMI


if ID  higher than IMPORT, plugin say:"line 1:0 missing 'import' at 'import'
line 1:7 extraneous input 'a' expecting {'.', ';'}"
result of program also mismatched input 'import' expecting 'import'.

**Reply by lexasub** [reply]
output of plugin(leafs)
IMPORT:missing 'import' ID:import ID:a SEMI

**Reply by lexasub** [reply]
i don't know, why plugin output != result of program. result of program is line 1:0 mismatched input 'import' expecting 'import', but plugin give tree without error

**Comment by lexasub** [comment]
as far as I understand, first you need to list complete matches in the lexical rules. Symbols and keywords. then we write lexical rules like ID. I tried to make ID a parser rule, but still the problem persists. in the plugin, everything is parsed normally (the tree is normally built), and when the program starts, the result is different.

**Reply by parrt** [reply]
you should get the same results. only explanation i've ever seen is when the grammars were different or there's a bug in the target language antlr lib you use to parse. or you've got an error in how you call or something.  You have identified the situation above:

> if ID higher than IMPORT, plugin say:"line 1:0 missing 'import' at 'import'

**Reply by lexasub** [reply]
my version is antlr4.10.1. I barely made up main with visitor, maybe I did something wrong. I need the code for visitor.
CharStream stream = CharStreams.fromFileName("./test", StandardCharsets.US_ASCII);
langosLexer lexer = new langosLexer(stream);
CommonTokenStream tokens = new CommonTokenStream(lexer);
langosWithoutSyntaxParser parser = new langosWithoutSyntaxParser(tokens);
MylangosWithoutSyntaxVisitor visitor = new MylangosWithoutSyntaxVisitor();
System.out.print(visitor.visitEntry_point(parser.entry_point()));

**Reply by lexasub** [reply]
my custom visitor is public class MylangosWithoutSyntaxVisitor implements langosWithoutSyntaxVisitor {}

**Reply by parrt** [reply]
Your code looks fine so it's got to be something to do with the order of the rules or whatever. I recommend just verifying that input symbols are matched as the correct token type.  There's no point in me identifying it for you because I guarantee this is the issue.

**Reply by lexasub** [reply]
[fixed]my problem was in incorrect using combination of somesing(i did some"random" changes and it working)

please, add guide for visitor using(and it). my example:
CharStream stream = CharStreams.fromFileName("./test", StandardCharsets.US_ASCII);//you may use antlr reader, as in other examples
langosWithoutSyntaxLexer lexer = new langosWithoutSyntaxLexer(stream);
CommonTokenStream tokens = new CommonTokenStream(lexer);
langosWithoutSyntaxParser parser = new langosWithoutSyntaxParser(tokens);
MylangosWithoutSyntaxVisitor visitor = new MylangosWithoutSyntaxVisitor();


in you visitor class:

public class MylangosWithoutSyntaxVisitor extends langosWithoutSyntaxBaseVisitor<String> {
@Override  public String visitProgram(langosWithoutSyntaxParser.ProgramContext ctx){
        if(ctx.import_() != null) return visitImport_(ctx.import_());
        return visitElement(ctx.element());
   }
 ....
}

**Reply by lexasub** [reply]
thank you very much
